src/mrpro/operators/models/TransientSteadyStateWithPreparation.py

```diff
@@ -0,0 +1,123 @@
+            inversion. During this time the magnetisation follows Mz(t) the signal model:
+                    Mz(t) = M0 + (m0_scaling_preparation*M0 - M0)e^(-t / T1)
+
+    Part C: After the spoiler gradient the data acquisition starts and the magnetisation Mz(t) can be described by the
+            signal model:
+                    Mz(t) = M0* + (M0_init - M0*)e^(-t / T1*)
+            where the initial magnetisation is
+                    M0_init = M0 + (m0_scaling_preparation*M0 - M0)e^(-delay_after_preparation / T1)
+            the effective longitudinal relaxation time is
+                    T1* = 1/(1/T1 - 1/repetition_time ln(cos(flip_angle)))
+            and the steady-state magnetisation is
+                    M0* = M0 T1* / T1
+
+    """
+
+    def __init__(
+        self,
+        sampling_time: float | torch.Tensor,
+        repetition_time: float,
+        m0_scaling_preparation: float = 1.0,
+        delay_after_preparation: float = 0.0,
+    ):
+        """Initialize transient steady state signal model.
+
+        Parameters
+        ----------
+        sampling_time
+            time points when model is evaluated
+            with shape (time, ...)
+        repetition_time
+            repetition time
+        m0_scaling_preparation
+            scaling of the equilibrium magnetisation due to the preparation pulse before the data acquisition
+        delay_after_preparation
+            Time between preparation pulse and start of data acquisition.
+            During this time standard longitudinal relaxation occurs.
+
+        """
+        super().__init__()
+        sampling_time = torch.as_tensor(sampling_time)
+        self.sampling_time = torch.nn.Parameter(sampling_time, requires_grad=sampling_time.requires_grad)
+        self.repetition_time = repetition_time
+        self.m0_scaling_preparation = m0_scaling_preparation
+        self.delay_after_preparation = delay_after_preparation
+
+    def forward(self, m0: torch.Tensor, t1: torch.Tensor, flip_angle: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Apply transient steady state signal model.
+
+        Parameters
+        ----------
+        m0
+            equilibrium signal / proton density
+            with shape (... other, coils, z, y, x)
+        t1
+            longitudinal relaxation time T1
+            with shape (... other, coils, z, y, x)
+        flip_angle
+            flip angle of data acquisition
+            with shape (... other, coils, z, y, x)
+
+        Returns
+        -------
+            signal
+            with shape (time ... other, coils, z, y, x)
+        """
+        delta_ndim = m0.ndim - (self.sampling_time.ndim - 1)  # -1 for time
+        sampling_time = self.sampling_time[..., *[None] * (delta_ndim)] if delta_ndim > 0 else self.sampling_time
+
+        # effect of preparation pulse
+        m_start = m0 * self.m0_scaling_preparation
+
+        # relaxation towards M0
+        t1 = torch.where(t1 == 0, 1e-10, t1)
```

Comment:
I remember @Stef-Martin reporting that using torch.where() might be problematic in some cases? What's the status here? Does @fzimmermann89 know something about this?

Reply:
Why is this required?

I would only ever add these hacks if in the end they result in the same value and are only required for numerical reasons.
It is arbitrary that the signal model returns something if t1 is exactly 0.

Reply:
if that is the case, we might add a comment

also, what about t1=1e-30?

Reply:
```
lncostr=torch.log(torch.cos(flip_angle)) / repetition_time
r1star=1/t1 - lncostr
m0star=m0/(1-t1*lncostr)
signal = m0_star + (m_start - m0_star) * torch.exp(-sampling_time*r1_star)
```